trainer_tf_barebones.py

```diff
@@ -0,0 +1,366 @@
+            per_replica_features, per_replica_labels = next(iterator)
+
+            return self._accumulate_gradients(per_replica_features, per_replica_labels)
+
+        while True:
+            try:
+                yield _accumulate_next()
+            except tf.errors.OutOfRangeError:
+                break
+
+    def _accumulate_gradients(self, per_replica_features, per_replica_labels):
+        """Accumulates the gradients across all the replica."""
+        per_replica_loss = self.strategy.experimental_run_v2(self._forward, args=(per_replica_features, per_replica_labels))
+
+        return self.strategy.reduce(tf.distribute.ReduceOp.MEAN, per_replica_loss, None)
+
+    def _forward(self, features, labels):
+        """Forwards a training example and accumulates the gradients."""
+        per_example_loss = self._run_model(features, labels, True)
+        loss = tf.nn.compute_average_loss(per_example_loss, global_batch_size=self.args.per_gpu_train_batch_size)
+        vars = self.model.trainable_variables
+        if self.args.mode == "labelling":
+            vars = [var for var in self.model.trainable_variables if "pooler" not in var.name]
+        gradients = self.optimizer.get_gradients(loss, vars)
+
+        self.gradient_accumulator(gradients)
+
+        return per_example_loss
+
+    def _run_model(self, features, labels, training):
+        """
+        Computes the loss of the given features and labels pair.
+        Args:
+          features: the batched features.
+          labels: the batched labels.
+        """
+        if self.args.mode == "classification" or self.args.mode == "labelling":
+            logits = self.model(features, training=training)[0]
+        else:
+            logits = self.model(features, training=training)
+
+        if self.args.mode == "labelling":
+            active_loss = tf.reshape(labels, (-1,)) != -1
+            logits = tf.boolean_mask(tf.reshape(logits, (-1, len(self.dataset_info.labels))), active_loss)
+            labels = tf.boolean_mask(tf.reshape(labels, (-1,)), active_loss)
+
+        loss = self.loss(labels, logits)
+
+        if training:
+            self.train_acc_metric(labels, logits)
+        else:
+            self.test_acc_metric(labels, logits)
+
+        return loss
+
+    def test(self) -> None:
+        """
+        Test the model over the test dataset and print a report.
+        """
+        y_true = []
+        results = self.model.predict(self.test_dataset, steps=self.test_steps)
+
+        if self.args.mode == "classification":
+            for batch in self.test_dataset:
+                y_true.extend(batch[1].numpy().tolist())
+
+            y_pred = np.reshape(np.argmax(results, axis=-1), (-1, 1)).tolist()
+            y_true = list(itertools.chain.from_iterable(y_true))
+            y_pred = list(itertools.chain.from_iterable(y_pred))
+
+            logger.info(classification_report(y_true, y_pred, target_names=self.dataset_info.labels))
```

Comment:
These two methods are the ones that are task dependent. The first one because of the loss computation and the second one because of the metrics computation. I was thinking, to remove this "dependency" to create a `losses_tf.py` and `metrics_tf.py` files where all the "custom" metrics and losses will be implemented and then becomes fully agnostic because these losses and metrics could be selected in the the training arguments. What do you think?